src/main/java/mesosphere/marathon/experimental/storage/PathTrie.java

```diff
@@ -0,0 +1,371 @@
+     * clear all nodes
+     */
+    public void clear() {
+        for(String child : rootNode.getChildren()) {
+            rootNode.deleteChild(child);
+        }
+    }
+
+    /**************************************************************************
+     *                           Additional methods                           *
+     **************************************************************************/
+
+    /**
+     * return trie's root node. Useful when iterating through the trie.
+     *
+     * @return root node
+     */
+    public TrieNode getRoot() {
+        return rootNode;
+    }
+
+    /**
+     * return a trie's node for the given path. If the path doesn't exist null is returned
+     *
+     * @param path input path
+     * @return node with the given path
+     */
+    TrieNode getNode(String path) {
+        if (path == null) {
+            return null;
+        }
+        if ("/".equals(path)) {
+            return rootNode;
+        }
+        String[] pathComponents = path.split("/");
+        if (pathComponents.length <= 1) {
+            throw new IllegalArgumentException("Invalid path " + path);
+        }
+        TrieNode parent = rootNode;
+        String part = null;
+        for (int i=1; i<pathComponents.length; i++) {
+            part = pathComponents[i];
+            if (parent.getChild(part) == null) {
+                //the path does not exist
+                return null;
+            }
+            parent = parent.getChild(part);
+            LOG.debug("{}",parent);
+        }
+        return parent;
+    }
+
+    /**
+     * returns node data if node exists or null otherwise
+     *
+     * @param path input path
+     * @return node data
+     */
+    public byte[] getNodeData(String path) {
+        TrieNode node = getNode(path);
+        if (node != null) {
+            return node.getData();
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * returns true if a node with a given path exists in the trie
+     *
+     * @param path input path
+     * @return true if the node exists, false otherwise
+     */
+    public Boolean existsNode(String path) {
+        TrieNode node = getNode(path);
+        return node != null;
+    }
+
+    private void findLeafNodes(String path, @NotNull TrieNode node, Set<String> leafs) {
+        if (node.isLeafNode()) {                 // If found leaf node (that is not root node)
+            leafs.add(path);                     // add it to the set
+        } else {                                 // else search all the children recursively
+            node.getChildrenNodes()
+                    .forEach((cp, n) ->
+                            findLeafNodes(Paths.get(path, cp).toString(), n, leafs));
+        }
+    }
+
+    /**
+     * return leaf nodes recursively starting with the given path.
+     *
+     * @param path input path
+     * @return a map with leaf nodes
+     */
+    public Set<String> getLeafs(String path) {
```

Comment:
Isn't this `findLeafNodes(path, getNode(path), new HashSet<>())`? I would call the the same.